src/common/io_util.cc

```diff
@@ -131,57 +97,66 @@ Status SockSetTcpKeepalive(int fd, int interval) {
 }
 
 Status SockConnect(const std::string &host, uint32_t port, int *fd, int conn_timeout, int timeout) {
-  if (conn_timeout == 0) {
-    auto s = SockConnect(host, port, fd);
-    if (!s) return s;
-  } else {
-    *fd = socket(AF_INET, SOCK_STREAM, 0);
-    if (*fd == NullFD) return Status::FromErrno();
-  }
+  addrinfo hints, *servinfo = nullptr, *p = nullptr;
 
-  auto exit = MakeScopeExit([fd] {
-    close(*fd);
-    *fd = NullFD;
-  });
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
 
-  if (conn_timeout != 0) {
-    sockaddr_in sin{};
-    sin.sin_family = AF_INET;
-    sin.sin_addr.s_addr = inet_addr(host.c_str());
-    sin.sin_port = htons(port);
+  if (int rv = getaddrinfo(host.c_str(), std::to_string(port).c_str(), &hints, &servinfo); rv != 0) {
+    return {Status::NotOK, gai_strerror(rv)};
+  }
+  auto exit = MakeScopeExit([servinfo] { freeaddrinfo(servinfo); });
 
-    fcntl(*fd, F_SETFL, O_NONBLOCK);
-    connect(*fd, reinterpret_cast<sockaddr *>(&sin), sizeof(sin));
+  for (p = servinfo; p != nullptr; p = p->ai_next) {
+    auto cfd = UniqueFD(socket(p->ai_family, p->ai_socktype, p->ai_protocol));
+    if (!cfd) continue;
 
-    auto retmask = Util::aeWait(*fd, AE_WRITABLE, conn_timeout);
-    if ((retmask & AE_WRITABLE) == 0 || (retmask & AE_ERROR) != 0 || (retmask & AE_HUP) != 0) {
-      return Status::FromErrno();
+    if (conn_timeout == 0) {
+      if (connect(*cfd, p->ai_addr, p->ai_addrlen) == -1) {
+        continue;
+      }
+    } else {
+      fcntl(*cfd, F_SETFL, O_NONBLOCK);
+      int ret = connect(*cfd, p->ai_addr, p->ai_addrlen);
+      if (ret != 0 && errno != EINPROGRESS) {
+        continue;
+      }
+      auto retmask = Util::aeWait(*cfd, AE_WRITABLE, conn_timeout);
+      if ((retmask & AE_WRITABLE) == 0 || (retmask & AE_ERROR) != 0 || (retmask & AE_HUP) != 0) {
+        return Status::FromErrno();
+      }
+
+      // restore to the block mode
+      int socket_arg = 0;
+      if ((socket_arg = fcntl(*cfd, F_GETFL, NULL)) < 0) {
+        return Status::FromErrno();
+      }
+      socket_arg &= (~O_NONBLOCK);
+      if (fcntl(*cfd, F_SETFL, socket_arg) < 0) {
+        return Status::FromErrno();
+      }
     }
 
-    int socket_arg = 0;
-    // Set to blocking mode again...
-    if ((socket_arg = fcntl(*fd, F_GETFL, NULL)) < 0) {
-      return Status::FromErrno();
+    Status s = SockSetTcpKeepalive(*cfd, 120);
+    if (s.IsOK()) {
+      s = SockSetTcpNoDelay(*cfd, 1);
     }
-    socket_arg &= (~O_NONBLOCK);
-    if (fcntl(*fd, F_SETFL, socket_arg) < 0) {
-      return Status::FromErrno();
+    if (!s.IsOK()) {
+      continue;
```

Comment:
continue or return error directly, which is better here?

Reply:
Looks better to continue if it has other addresses to have a try.